File: src/main/java/com/github/cannor147/request/colorization/ColorizationScheme.java
package com.github.cannor147.request.colorization;

import com.github.cannor147.model.Color;
import com.github.cannor147.model.Territory;
import com.github.cannor147.painter.RGBColor;

import java.util.Map;

public abstract class ColorizationScheme {
    protected Color defaultColor = Color.SILVER;

    public Color getDefaultColor() {
        return defaultColor;
    }

    public void registerDefaultColor(Color defaultColor) {
        this.defaultColor = defaultColor;
    }

    public void prepareForCalculation(Map<Territory, ColorizationParameter> territoryToParameterMap) {
        // No operations.
    }

    public abstract RGBColor calculateColor(ColorizationParameter colorizationParameter);
}

File: src/main/java/com/github/cannor147/request/colorization/ColorizationScheme.kt
package com.github.cannor147.request.colorization

import com.github.cannor147.model.Color
import com.github.cannor147.model.Territory
import com.github.cannor147.painter.RGBColor
import java.util.*
import java.util.function.ToDoubleFunction

abstract class ColorizationScheme {
    protected var defaultColor = Color.SILVER

    fun registerDefaultColor(defaultColor: Color) {
        this.defaultColor = defaultColor
    }

    open fun prepareForCalculation(territoryToParameterMap: Map<Territory, ColorizationParameter>) {
        // No operations.
    }

    abstract fun calculateColor(colorizationParameter: ColorizationParameter): RGBColor
}

fun <T> Sequence<T>.summarizingDouble(toDoubleFunction: ToDoubleFunction<T>): DoubleSummaryStatistics {
    return this.fold(DoubleSummaryStatistics()) { statistics: DoubleSummaryStatistics, element: T ->
        statistics.accept(toDoubleFunction.applyAsDouble(element))
        statistics
    }
}